core/server/common/src/main/java/alluxio/master/AbstractMaster.java

```diff
@@ -313,4 +323,87 @@ public void close() {
       }
     }
   }
+
+  @NotThreadSafe
+  public final class MasterAuditContext implements AuditContext {
+    private final AsyncUserAccessAuditLogWriter mAsyncAuditLogWriter;
+    private boolean mAllowed;
+    private String mCommand;
+    private String mSrcPath;
+    private String mDstPath;
+    private String mUser;
+    private String mIp;
+    private String mSrcPathOwner;
+    private String mSrcPathGroup;
+    private short mSrcPathMode;
+
+    @Override
+    public MasterAuditContext setAllowed(boolean allowed) {
+      mAllowed = allowed;
+      return this;
+    }
+
+    public MasterAuditContext setCommand(String command) {
+      mCommand = command;
+      return this;
+    }
+
+    public MasterAuditContext setSrcPath(String srcPath) {
+      mSrcPath = srcPath;
+      return this;
+    }
+
+    public MasterAuditContext setDstPath(String dstPath) {
+      mDstPath = dstPath;
+      return this;
+    }
+
+    public MasterAuditContext setUser(String user) {
+      mUser = user;
+      return this;
+    }
+
+    public MasterAuditContext setIp(String ip) {
+      mIp = ip;
+      return this;
+    }
+
+    public MasterAuditContext setSrcPathOwner(String owner) {
+      mSrcPathOwner = owner;
+      return this;
+    }
+
+    public MasterAuditContext setSrcPathGroup(String group) {
+      mSrcPathGroup = group;
+      return this;
+    }
+
+    public MasterAuditContext setSrcPathMode(short mode) {
+      mSrcPathMode = mode;
+      return this;
+    }
+
+    private MasterAuditContext(AsyncUserAccessAuditLogWriter asyncAuditLogWriter) {
+      mAsyncAuditLogWriter = asyncAuditLogWriter;
+      mAllowed = true;
+    }
+
+    @Override
+    public void append() {
+      if (mAsyncAuditLogWriter != null) {
+        mAsyncAuditLogWriter.append(this);
+      }
+    }
+
+    @Override
+    public void close() {}
```

Comment:
I find it confusing too. Originally we let the same thread perform append and flush operation. Then I had a discussion with Calvin and now we have a separate thread perform audit context flush. Therefore, when the thread appending the audit context leaves the try-with-resources block's scope, it does not have to do anything. Does this make sense to you?